### PR TITLE
[MXNET-936] [WIP] [DO NOT REVIEW] Support projection and clip in CuDNN LSTM

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Support new features in recent CuDNN 7.1.1 and 7.2.1 releases.

### Essentials ###
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Support projection layer
- [ ] Support clip of hidden layer and projection
- [ ] Unit tests

### CODE REVIEW

File: python/mxnet/gluon/rnn/rnn_layer.py [hunk 1]
```diff
@@ -351,6 +353,8 @@ class LSTM(_RNNLayer):
     ----------
     hidden_size: int
         The number of features in the hidden state h.
+    projection_size: int, default None
+        The numer opf features after projection.
```
Comment (szha) [inline python/mxnet/gluon/rnn/rnn_layer.py]: typo

File: src/operator/cudnn_rnn-inl.h [hunk 2]
```diff
@@ -498,6 +498,16 @@ class CuDNNRNNOp : public Operator{
...
         CUDNN_CALL(cudnnSetRNNMatrixMathType(rnn_desc_, math_type));
       #endif
+      if (param_.projection_size.has_value()) {
+      #if (CUDNN_MAJOR == 7 && CUDNN_MINOR >= 1) || CUDNN_MAJOR > 7
+        CUDNN_CALL(cudnnSetRNNProjectionLayers(s->dnn_handle_,
+                                               rnn_desc_,
+                                               param_.projection_size.value(),
+                                               0));
+      #else
+        LOG(FATAL) << "Projection is only available with later than CuDNN 7.1.1";
```
Comment (szha) [inline src/operator/cudnn_rnn-inl.h]: does the projection work for all RNN types? if so, add the support to all RNN layers. if not, could you check for the type here and make sure it's supported?

Comment (haojin2) [reply src/operator/cudnn_rnn-inl.h]: Okay

**Comment by kalyc** [comment]
Thanks for your contribution @haojin2 
@mxnet-label-bot[pr-work-in-progress]

**Comment by szha** [comment]
When creating begin_states, the currently gluon RNN layer logic doesn't handle the dtype correctly (reported by @Ishitori). example:
```
fake_data = nd.random.uniform(shape=(batch_size, context_max_length, 8 * embedding_size), dtype="float16")
attention_output = nd.transpose(fake_data, axes=(1, 0, 2))
modeling_layer = LSTM(hidden_size=100, num_layers=2, dropout=0.2, bidirectional=True)
modeling_layer.cast("float16")
modeling_layer.initialize()
modeling_layer(attention_output)
```

@haojin2 could you include the fix?

**Pull request closed without merging**